File: java/base/src/main/vsdk/toolkit/environment/light/DirectionalLight.java
package vsdk.toolkit.environment.light;

import vsdk.toolkit.common.linealAlgebra.Vector3Dd;
import vsdk.toolkit.common.color.ColorRgb;
import vsdk.toolkit.environment.geometry.element.Ray;

public final class DirectionalLight extends Light
{
    public DirectionalLight(Vector3Dd direction, ColorRgb emission)
    {
        super(direction.normalized(), emission);
    }

    @Override
    public LightDirection getDirectionAndDistance(Vector3Dd surfacePoint)
    {
        Vector3Dd direction = getPosition();
        return new LightDirection(
            new Vector3Dd(-direction.x(), -direction.y(), -direction.z()),
            Double.POSITIVE_INFINITY);
    }

    @Override
    public double evaluateLightResponseFactor(Ray lightSourceRay)
    {
        return 1.0;
    }

    @Override
    public Light copy()
    {
        DirectionalLight copy = new DirectionalLight(getPosition(), getEmission());
        copy.setId(getId());
        copy.setName(getName());
        return copy;
    }
}

File: typescript/base/src/main/vsdk/toolkit/environment/light/DirectionalLight.ts
import { ColorRgb } from "../../common/color/ColorRgb.js";
import { Vector3Dd } from "../../common/linealAlgebra/Vector3Dd.js";
import type { Ray } from "../geometry/element/Ray.js";
import { Light } from "./Light.js";
export class DirectionalLight extends Light {
    public constructor(direction: Vector3Dd, emission: ColorRgb) {
        super(direction.normalized(), emission);
    }
    public override getDirectionAndDistance(_point: Vector3Dd): Light.LightDirection {
        const d = this.getPosition();
        return new Light.LightDirection(new Vector3Dd(-d.x(), -d.y(), -d.z()), Infinity);
    }
    public override evaluateLightResponseFactor(_ray: Ray): number {
        return 1;
    }
    public override copy(): Light {
        const copy = new DirectionalLight(this.getPosition(), this.getEmission());
        copy.setId(this.getId());
        copy.setName(this.getName());
        return copy;
    }
}
